chain/pool/src/lib.rs

```diff
@@ -118,6 +111,80 @@ impl TransactionPool {
     }
 }
 
+pub struct DrainingIterator<'a> {
+    pool: &'a mut TransactionPool,
+    sorted: bool,
+    current_map: Cell<TxMap>,
+    next_map: Cell<TxMap>,
+    last_entry: Option<((AccountId, PublicKey), Vec<SignedTransaction>)>,
+}
+
+impl<'a> DrainingIterator<'a> {
+    pub fn new(pool: &'a mut TransactionPool) -> Self {
+        let current_map = Default::default();
+        pool.transactions.swap(&current_map);
+        Self { pool, sorted: false, current_map, next_map: Default::default(), last_entry: None }
+    }
+
+    /// Returns the next transaction in the proper order. `same_key` defines whether the
+    /// transaction should be from the same group (with the same key) as the previous transaction.
+    /// If the previous transaction was invalid, the next transaction has to be with the same key to
+    /// maintain the proper order. Otherwise the invalid transaction skips the given key.
+    fn next(&mut self, same_key: bool) -> Option<SignedTransaction> {
+        // If we need the new/next key, or the current transaction group is fully used.
+        if !same_key || self.last_entry.is_none() {
+            if let Some((key, txs)) = self.last_entry.take() {
+                self.next_map.get_mut().insert(key, txs);
+            }
+            if self.current_map.get_mut().is_empty() {
+                self.sorted = true;
+                self.current_map.swap(&self.next_map);
+            }
+            let key = if let Some(key) = self.current_map.get_mut().keys().next() {
+                key.clone()
+            } else {
+                return None;
+            };
+            self.last_entry = self.current_map.get_mut().remove_entry(&key);
+            if !self.sorted {
+                // Sort by nonce in non-increasing order to pop from the end
+                self.last_entry
+                    .as_mut()
+                    .unwrap()
+                    .1
+                    .sort_by_key(|a| std::cmp::Reverse(a.transaction.nonce));
+            }
+        }
+
+        if self.last_entry.is_some() {
+            let tx = self
+                .last_entry
+                .as_mut()
+                .unwrap()
+                .1
+                .pop()
+                .expect("transaction groups shouldn't be empty");
+            if self.last_entry.as_ref().unwrap().1.is_empty() {
+                self.last_entry = None;
+            }
+            self.pool.unique_transactions.remove(&tx.get_hash());
+            Some(tx)
+        } else {
+            None
+        }
+    }
+}
+
+impl<'a> Drop for DrainingIterator<'a> {
+    fn drop(&mut self) {
+        if let Some((key, txs)) = self.last_entry.take() {
+            self.current_map.get_mut().insert(key, txs);
+        }
+        self.pool.transactions.swap(&self.current_map);
```

Comment:
We have mutable reference to `self` so I don't think `Cell` is useful here. Could use deconstruction to swap it:
```rust
let Self{pool, current_map,..} = self;
std::mem::swap(&mut pool.transactions, current_map);
```
Swapping, however, is probably reallocating the two objects, both in `Cell` and `mem::swap` cases, these might not be a problem if transactions is a collections, since it is a shallow move

Reply:
Replaced with a helper function. Don't move maps at all anymore